Planner: improve resilience against fluctuations and short charging slots

schenlap opened the pull request.

schenlap committed:
fix charge time correction for high target soc

schenlap committed:
ignore all short charging slots

schenlap committed:
continue during short off periods

schenlap commented:
ist mal zur Diskussion

andig commented on core/loadpoint.go:
```diff
@@ -817,6 +815,9 @@ func (lp *Loadpoint) plannerActive() (active bool) {
 		case requiredDuration < 30*time.Minute:
 			lp.log.DEBUG.Printf("continuing for remaining %v", requiredDuration.Round(time.Second))
 			active = true
+		case lp.clock.Now().Add(10*time.Minute).After(planStart):
```
lp.clock.Until() < xx

andig commented on core/loadpoint.go:
```diff
@@ -794,7 +792,7 @@ func (lp *Loadpoint) plannerActive() (active bool) {
 	if active {
 		// ignore short plans if not already active
-		if !lp.planActive && requiredDuration < 10*time.Minute {
+		if !lp.planActive && lp.clock.Now().Add(10*time.Minute).After(slotEnd) {
```
lp.clock.Until() < xx

andig replied:
10min als const:

    smallSlotDuration = 10 * time.Minute // small planner slot duration we might ignore

andig commented on core/loadpoint.go:
```diff
@@ -775,12 +775,10 @@ func (lp *Loadpoint) plannerActive() (active bool) {
-			requiredDuration = time.Duration(float64(requiredDuration) / soc.ChargeEfficiency)
-		}
+	if targetSoc > 80 && maxPower > 4000 {
+		additionalTime := time.Duration(float64(targetSoc - 80) / (float64(targetSoc) - lp.vehicleSoc) * float64(requiredDuration) * (1 - soc.ChargeEfficiency))
```
Ist die Frage, ob das reicht hier 1x Effizienz draufzuschlagen?

schenlap replied:
nach https://app.slack.com/client/T013D811Q1E/browse-files/rimeto_profile/U02RN534WSG braucht der Tesla ungefähr 15min länger  weil ab ~94% reduziert wird. Aktuell wird hier 10min draufgeschlagen, das ist dann zu wenig. Sollten wir auch erst später reduzieren (90%) und dafür stärker?
Eigentlich müsste man bei >12kW ab 80%, bei >4kw ab 90%. Aber wollen wir es so kompliziert?

Ein paar gefundene Referenzen:
https://uploads.tff-forum.de/optimized/3X/4/d/4db9a5f4fbc6367fd86b3df58648ebd03f1f011c_2_690x344.png
https://i.imgur.com/aho51Fh.png

Hofyyy replied:
Anbei mal die Ladekurve von meinem Tesla für 11KW (wenn es nicht -10 Grad sind). Das kann sich natürlich bei verschiedenen Akkus (LPF + NMC) weiter unterscheiden. Aber so haben wir mal einen weiteren Anhaltspunkt. 
<img width="625" alt="ladekurve" src="https://user-images.githubusercontent.com/68365998/210978414-ed7ab0b8-1c25-4d02-ac71-df36aa52ea06.png">

schenlap commented:
@andig da ist jetzt leider was schief gegangen und deine Commit 0cbd1869e9d341da2946b7e503c0481f47b6809c  simplify wurde überschrieben. Sorry, ich repariere das.

andig commented on core/loadpoint.go:
```diff
@@ -817,6 +815,9 @@ func (lp *Loadpoint) plannerActive() (active bool) {
 		case requiredDuration < 30*time.Minute:
 			lp.log.DEBUG.Printf("continuing for remaining %v", requiredDuration.Round(time.Second))
 			active = true
+		case lp.clock.Now().Add(10 * time.Minute).After(planStart):
```
dito

schenlap committed:
simplify

Hofyyy commented:
Wollen wir zum Einstellen der Verlustleistung einen eigenen PR aufmachen? Hier scheint es ja eher um die Ladekurve am Ende und kleineren Fixes zu gehen?

Also Default 10%, aber ggf. pro Auto einstellbar, da es sich ja doch sehr unterscheidet pro Hersteller und Setup.

schenlap committed:
better chargetime calculation

calc extra charge time for target 100% and soc < 80%:
27min @ 22kW
16min @ 11kW (for soc < 90%)
0min  @ 3.6kW

schenlap committed:
coding style

Hofyyy commented:
Da es gerade in einem anderen Thread parallel ging:
- Ab 90% reicht sehr wahrscheinlich Faktor 2 und nicht Faktor 3
- Ab 80% (meine Meinung) brauchen wir denke ich nicht, da die DC Schnelladekurven meistens von 80% -> 90% von 50KWH auf 25KWH reduzieren. Da wir aber sowieso nur bis 22KW am AC laden (annahme) brauchen wir da garnichts reduzieren. Das wäre anders, wenn wir jetzt zwei Steps brauchen. z.B. 90-95 und dann noch mal 95-100

premultiply commented:
Ich würde diese ganze Ladekurven-Thematik und die daraus von verschiedenen Faktoren abhängende Zeitschätzung gerne vollständig in den Estimator verlagern.

schenlap commented:
Ja, das gehört schlussendlich in den Estimator. Jetzt geht es mal darum zu wissen was da sinnvoll ist.

schenlap commented:
> Die halten alle bis 90% mindestens 22KW

Der SmartEQ reduziert jedenfalls bei ca. 82% von 22kW auf 11kW.

premultiply commented:
Lass uns das hier erstmal so grob lassen in dem Wissen dass es später nochmal grundsätzlich verbessert wird.
Ich habe da schon eine recht konkrete Idee.

schenlap commented on core/loadpoint.go:
```diff
@@ -817,6 +822,9 @@ func (lp *Loadpoint) plannerActive() (active bool) {
 		case requiredDuration < 30*time.Minute:
 			lp.log.DEBUG.Printf("continuing for remaining %v", requiredDuration.Round(time.Second))
 			active = true
+		case lp.clock.Until(planStart) < smallSlotDuration:
+			lp.log.DEBUG.Printf("plan will re-start shortly, continuing for remaining %v", lp.clock.Until(planStart).Round(time.Second))
+			active = true
```
Sollten wir nicht auch wenn wir länger eingeschalten lassen als der Planer vorgegeben hat dies speichern, also immer 
```
		lp.setPlanActive(active)
		lp.planSlotEnd = slotEnd
```
machen? Ich hatte jetzt den fall dass zuerst  `plan will re-start shortly, continuing for remaining` und das langsam auf 10 Minuten hochgezählt hat und dann abgeschaltet wurde. (Auto hat schneller geladen als gedacht). Eigentlich hätte ich erwartet dass dann `continuing until end of slot` einspringt, ist aber nicht passiert weil dieser Zustand nicht gespeichert wird. @andig Spricht was dagegend setPlanActive(active) zu setzen vor dem return. Was ist der Grund warum die Änderungen nicht übernommen werden?

andig replied:
> @andig Spricht was dagegend setPlanActive(active) zu setzen vor dem return. Was ist der Grund warum die Änderungen nicht übernommen werden?

@schenlap könnten wir leicht im `defer` machen falls active? Nebenwirkungen?

schenlap replied:
OK. Defer war nicht notwendig, ist nur an einer Stelle. Nebenwirkungen: schau ma mal...

schenlap committed:
save overriden planActive state

andig commented on core/loadpoint.go:
```diff
@@ -817,9 +818,15 @@ func (lp *Loadpoint) plannerActive() (active bool) {
 
+	// remember last active plan's end time
+	lp.setPlanActive(active)
+	lp.planSlotEnd = slotEnd
```
machst du damit im fall !active nicht das `lp.slotEnd` kaputt? Denke das muss oben bleiben.

schenlap replied:
done

schenlap committed:
fix last active plan's end time

schenlap commented:
![image](https://user-images.githubusercontent.com/20611587/211142261-07f6d89c-c71c-4027-8ad7-239dcd2d3a5b.png)
schaut gut aus. Auto wurde aber nicht voll aufgeladen, dazu kann ich nichts sagen.

andig commented:
Wenn nicht voll hätte er egtl. wieder anfangen müssen bzw. erst gar nicht aufhören. Magst du nochmal ins Log schauen?

schenlap commented:
Nö, da habe ich mich falsch ausgedrückt. Er hat korrekt geladen, ich hatte den ZielSoC nicht auf 100%. Planner hat alles richtig gemacht

schenlap commented:
Es gibt noch das Problem dass ein alter Plan beim anstecken wieder genommen wird.
Auto anstecken, mit Planner laden -> Laden beendet -> Auto abstecken -> fahren -> Auto anstecken und die ALTE Zielzeit wird wieder hergenommen. Wenn notwendig startet das Auto auch sofort mit dem Laden.

Es wird der alte Planstart im UI noch angezeigt.

@andig 
Ich denke wir müssen die Zielzeit löschen wenn das Auto abgesteckt wurde? Ich weiß aber nicht an welcher Stelle das am Besten passieren soll.

![image](https://user-images.githubusercontent.com/20611587/211144423-c6912e48-a1a0-452f-9905-3644dde1b927.png)

schenlap committed:
fix planactive in off and disconnected mode

schenlap commented:
@andig 
Das wäre fertig, ich bin damit zufrieden. Tuning bedarf es sicher noch bei additionalTime > 80 / 90%, da bekommen wir dann erst Feedback wenn's im nightly ist, eine Verbesserung zum jetztigen Zustand ist es aber sicher.

Offen ist auch noch dass die TargetTime gelöscht wird, z.B wenn Fahrzeug abgesteckt wird. Soll ich dazu ein Issue öffnen?

andig commented:
Bitte separate Diskussion- gerade wenn wir Schedules einführen (jeden Werktag...) dann wollen wir da ja nix löschen.

andig committed:
Move timer resets to mode switch

andig committed:
Consolidate returns

andig commented:
@schenlap schau bitte nochmal die letzten 2 Commits an- ok?

schenlap commented on core/loadpoint_api.go:
```diff
@@ -43,8 +43,15 @@ func (lp *Loadpoint) SetMode(mode api.ChargeMode) {
+		case api.ModeNow, api.ModeOff:
+			lp.resetPhaseTimer()
+			lp.resetPVTimer()
+			lp.setPlanActive(false)
```
für Mode Now ist es jetzt neu. Kann man aber so lassen.

schenlap commented on core/loadpoint.go:
```diff
@@ -752,6 +752,7 @@ func (lp *Loadpoint) setPlanActive(active bool) {
 // plannerActive checks if charging plan is active
 func (lp *Loadpoint) plannerActive() (active bool) {
 	defer func() {
+		lp.setPlanActive(active)
```
active wird aber unten nicht gesetzt, also hilft das defer nicht. Ich brauche auch die überschriebenen Zustände. Ein ist Ein, egal wer schlussendlich eingeschaltet hatte (planner oder hier im loadpoint)

andig replied:
active ist die Returnvariable- und die wird vom `return` gesetzt ;)

schenlap replied:
:+1:

andig commented on core/loadpoint.go:
```diff
@@ -1848,6 +1856,7 @@ func (lp *Loadpoint) Update(sitePower float64, batteryBuffered bool) {
 		// always disable charger if not connected
 		// https://github.com/evcc-io/evcc/issues/105
 		err = lp.setLimit(0, false)
+		lp.setPlanActive(false)
```
Der kann doch auch weg- ist ja schon im Disconnect Handler?

andig committed:
wip
